File: 147.php
<?php

/**
 * Definition for a singly-linked list.
 * class ListNode {
 *     public $val = 0;
 *     public $next = null;
 *     function __construct($val) { $this->val = $val; }
 * }
 */
class ListNode {
  public $val = 0;
  public $next = null;

  function __construct($val) {
    $this->val = $val;
  }
}

class Solution {

  /**
   * @param ListNode $head
   * @return ListNode
   */
  function insertionSortList($head) {
    if (!$head || !$head->next) {
      return $head;
    }

    $dummy = new ListNode(null);
    $dummy->next = $head;
    while ($head && $head->next) {
      if ($head->val < $head->next->val) {
        $head = $head->next;
        continue;
      }

      $pre = $dummy;
      while ($pre->next->val < $head->next->val) {
        $pre = $pre->next;
      }
      $next = $head->next;
      $head->next = $next->next;
      $next->next = $pre->next;
      $pre->next = $next;

      var_dump($dummy->next);
    }
    return $dummy->next;
  }
}

$node1 = new ListNode(4);
$node2 = new ListNode(2);
$node3 = new ListNode(1);
$node4 = new ListNode(3);
$node1->next = $node2;
$node2->next = $node3;
$node3->next = $node4;

$solution = new Solution();
var_dump($solution->insertionSortList($node1));

File: go/linkedList/147.go
package main

//func insertionSortList(head *ListNode) *ListNode {
//	if head == nil || head.Next == nil {
//		return head
//	}
//	dummy := new(ListNode)
//	dummy.Next = head
//
//	for head.Next != nil {
//		if head.Val < head.Next.Val {
//			head = head.Next
//		} else {
//			cur := dummy
//			for cur.Next.Val < head.Next.Val {
//				cur = cur.Next
//			}
//
//			next := head.Next
//			head.Next, next.Next, cur.Next = next.Next, cur.Next, next
//		}
//	}
//
//	return dummy.Next
//}
//
//type ListNode struct {
//	Val  int
//	Next *ListNode
//}
//
//func main() {
//	node1 := &ListNode{Val: 4}
//	node2 := &ListNode{Val: 3}
//	node3 := &ListNode{Val: 2}
//	node4 := &ListNode{Val: 1}
//
//	node1.Next = node2
//	node2.Next = node3
//	node3.Next = node4
//
//	insertionSortList(node1)
//}
